### DISCUSSION TITLE
Disable automatic fetch after mutations

### DISCUSSION THREAD

**Discussion by ragokan** [urql-graphql/urql#2259] [Q&A] [answered]
Hello, currently I have a query and a mutation to test the URQL.  For now, it looks amazing, but I want to learn how to use a feature. Basically I want to disable automatic re-fetch after mutations.

This is my code:
```ts
const [{ data: todos }] = useQuery(getTodosDocument);
const [result, addTodo] = useMutation(addTodoDocument);
```
This is my client's *cacheExchange*:
```ts
cacheExchange({
 schema: introspectedSchema,
 updates: {
    Mutation: {
      addTodo(result, _args, cache, _info) {
        cache.updateQuery({ query: getTodosDocument }, (data) => {
          data?.todos.push(result.todo);
          return data;
        });
// braces here
```

This example disable auto re-fetching when *todos* is an empty array. But if it is not it re-calls the *useQuery* of *getTodosDocument*. There is my server logs, as you can see it re-calls the *getTodos query* every time a mutation happens.
```
addTodo mutation
getTodos query
addTodo mutation
getTodos query
addTodo mutation
getTodos query
```

**Comment by JoviDeCroock** [comment, marked as answer]
That will most likely be because the todo you are adding is missing a required property or you are adding null to the array.

Basically schema-awareness enables nullability detection so if we see missing properties in the todo you added or a non-nullable null a refetch will occur.

Found the cause

          data?.todos.push(result.addTodo);
          
 You are pushing `undefined` on the array because it's `result.addTodo `

**Reply by ragokan** [reply]
I don't think that it is the case. 
data -> { todos: array of todos }
result.todo -> new todo
result.addTodo -> new todo, again, as same as todo (probably because I did todo: addTodo in mutation)

Length of data?.todos before push -> 5
after push -> 6.

I tried logging all of them individually, look like there is no problem.

**Reply by ragokan** [reply]
Btw, may I also disable it with code? Like: `"autoInvalidate" : false`

**Reply by JoviDeCroock** [reply]
I mean, could you reproduce it in a sandbox 😅 refetching _only_ occurs when there's something wrong

Which you can verify in our examples https://github.com/FormidableLabs/urql/tree/main/examples/with-graphcache-updates

**Reply by ragokan** [reply]
Hello, I don't know how can I upload the server but let me firstly show this video while I am trying to upload the server to somewhere.

https://vimeo.com/674547429

(now uploading the server)

**Reply by JoviDeCroock** [reply]
I mean, just using a codesandbox or logging out the updater would be sufficient 😅 the reason why would probably even show itself when you show us the queries and log out the mutation response

**Reply by JoviDeCroock** [reply]
https://codesandbox.io/s/wizardly-panna-y2moe?file=/src/components/Todos.js Here's an example

**Reply by JoviDeCroock** [reply]
The issue here @ragokan is that you are duplicating all exchanges and work with both doc and norm cache. This because the `defaultExchanges` export already has a cache. The example I linked above + our examples and docs 

Show this

https://github.com/FormidableLabs/urql/blob/main/examples/with-graphcache-pagination/src/App.jsx#L2-L21



```js
import { createClient, dedupExchange, fetchExchange } from "urql";

const client = createClient({
  url: "your-url",
  exchanges: [
    dedupExchange,
    cacheExchange({
      updates: {
        Mutation: {
          addTodo: (result, _args, cache) => {
            cache.updateQuery({ query: getTodosDocument }, (data) => {
              console.log(data, result);
              data.todos.push(result.addTodo);
              return data;
            });
          }
        }
      }
    }),
    fetchExchange
  ]
});
```

**Reply by ragokan** [reply]
Ahh, so when I add both *fetchExchange* and *dedupExchange* the problem is solved. Thank you!

**Reply by ragokan** [reply]
Lastly, I want to ask one more question. How can I disable automatic updates? Lets say that I really don't want to re-fetch the query?

**Reply by JoviDeCroock** [reply]
Again the normalized cache supports default data-semantics, we only refetch when it's actually needed like if you were to push a todo on your array that does not adhere to the selection-set of your query. Refetching in these scenario's makes it so your UI does not go into indeterministic states, being able to disable that would be a footgun for all users 😅

**Reply by ragokan** [reply]
> _Again_ the normalized cache supports default data-semantics, we only refetch when it's actually needed like if you were to push a todo on your array that does not adhere to the selection-set of your query. Refetching in these scenario's makes it so your UI does not go into indeterministic states, being able to disable that would be a footgun for all users 😅

Ahh, I see. Thank you so much for your answers. Still if there is any way to disable I would like to learn :D
Btw, I got my answer and my problem is gone, if you want you can close this discussion (I don't know how to close).

**Reply by kitten** [reply]
Mate, please read the docs, it's all in there 😅 https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/

**Comment by ragokan** [comment]
@JoviDeCroock hello. Please try it here: https://codesandbox.io/s/youthful-jennings-cddws?file=/src/App.tsx

To watch the logs, you can use: https://ragotodosapp.herokuapp.com/logs

I am logging the queries. As you can see it calls the get function every time a mutation happens.

**Comment by ragokan** [comment]
I don't think the problem is backend but I included all the backend code below.

<details> 
<summary>To see backend code (module) click</summary>

```ts

import { createApplication, createModule, gql } from "graphql-modules";

let id: number = 0;
const todos = [];

export const logs = [];

const defaultResolvers = {
  Query: {
    getTodos: () => {
      console.log("get todos query");
      logs.push("get todos query");
      return todos;
    },
  },
  Mutation: {
    addTodo: (_, { title }) => {
      console.log("add todo mutation");
      logs.push("add todo mutation");
      const newTodo = { id: String(++id), title };
      todos.push(newTodo);
      return newTodo;
    },
  },
};

const defaultModule = createModule({
  id: "default",
  dirname: __dirname,
  typeDefs: gql`
    type Todo {
      id: ID!
      title: String!
    }

    type Query {
      getTodos: [Todo!]!
    }

    type Mutation {
      addTodo(title: String!): Todo!
    }
  `,

  resolvers: [defaultResolvers],
});

export const appModule = createApplication({
  modules: [defaultModule],
});

```
</details>

<details> 
<summary>To see backend code (app) click</summary>

```ts

import { envelop } from "@envelop/core";
import { useGraphQLModules } from "@envelop/graphql-modules";
import cors from "cors";
import express from "express";
import { getGraphQLParameters, processRequest, renderGraphiQL, sendResult, shouldRenderGraphiQL } from "graphql-helix";
import { appModule, logs } from "./modules/application";

const getEnveloped = envelop({
  plugins: [useGraphQLModules(appModule)],
});

const setup = async () => {
  const app = express();

  app.use(express.json());
  app.use(
    cors({
      origin: "*",
    })
  );

  app.set("trust proxy", 1);

  app.use("/graphql", async (req, res) => {
    const { parse, validate, contextFactory, execute, schema, subscribe } = getEnveloped({});
    const request = {
      body: req.body,
      headers: req.headers,
      method: req.method,
      query: req.query,
    };

    const { operationName, query, variables } = getGraphQLParameters(request);

    if (shouldRenderGraphiQL(request)) {
      return res.send(renderGraphiQL({ subscriptionsEndpoint: "ws://localhost:8000/graphql" }));
    }

    const result = await processRequest({
      operationName,
      query,
      variables,
      request,
      schema,
      parse,
      validate,
      execute,
      contextFactory,
      subscribe,
    });
    sendResult(result, res);
  });

  app.get("/logs", (_, res) => {
    res.status(200).json({ logs });
  });

  app.get("/", (_, res) => res.status(200).json({ message: "ok" }));

  app.listen(process.env.PORT ?? 8000, () => console.log("🚀 Server is running"));
};

setup();


```
</details>
